File: api/crs/codegen/lib/crs_solve_triut/crs_solve_triut.c
#include "crs_solve_triut.h"
#include "crs_solve_triut_types.h"

void crs_solve_triut(const CRS_Matrix *A, emxArray_real_T *b)
{
  int b_i;
  int i;
  int i1;
  int i2;
  int i3;
  int k;
  i = A->row_ptr->size[0];
  for (b_i = 0; b_i <= i - 2; b_i++) {
    b->data[b_i] /= A->val->data[A->row_ptr->data[b_i] - 1];
    i1 = A->row_ptr->data[b_i] + 1;
    i2 = A->row_ptr->data[b_i + 1] - 1;
    for (k = i1; k <= i2; k++) {
      i3 = A->col_ind->data[k - 1] - 1;
      b->data[i3] -= A->val->data[k - 1] * b->data[b_i];
    }
  }
}

void crs_solve_triut_initialize(void)
{
}

void crs_solve_triut_terminate(void)
{
}

File: api/crs/crs_solve_triut.m
function b = crs_solve_triut(A, b)
% crs_solve_triut Solves A.'\b, where A is upper triangular
%     b = crs_solve_triut(A, b)
%  The matrix is assumed to be upper triangular. The right-hand side vector
%  will be overwritten by A.'\b at output
%
% See also crs_solve_tril, crs_solve_triut

%#codegen -args {crs_matrix, coder.typeof(0, [inf, 1])}

coder.inline('always');

for i = 1:int32(length(A.row_ptr)) - 1
    cstart = A.row_ptr(i);
    assert(A.col_ind(cstart) == i && A.val(cstart) ~= 0);

    b(i) = b(i) / A.val(cstart);

    for k = (cstart + 1):(A.row_ptr(i+1) - 1)
        j = A.col_ind(k);
        % This can be done concurrently
        b(j) = b(j) - A.val(k) * b(i);
    end
end

function test %#ok<DEFNU>
%!test
%! for k=1:100
%!     R = triu(rand(10,10)); b = rand(10, 1);
%!     if condest(R) < 1.e5
%!         A = crs_createFromSparse(sparse(R));
%!         us = crs_solve_triut(A, b);
%!         assert(norm(R' * us - b) <= 1.e-10);
%!     end
%! end
